src/webgpu/util/conversion.ts

```diff
@@ -1112,6 +1112,35 @@ export function scalarTypeOf(ty: Type): ScalarType {
   throw new Error(`unhandled type ${ty}`);
 }
 
+/** @returns the concrete type of the given Type.
+ * Example: vec3<abstract-float> -> vec3<float>
+ */
+export function concreteTypeOf(ty: Type): Type {
+  let cty = elementTypeOf(ty);
+  if (cty === Type.abstractInt) {
+    cty = Type.i32;
+  } else if (cty === Type.abstractFloat) {
+    cty = Type.f32;
+  }
+
+  if (ty instanceof ScalarType) {
+    return cty;
+  }
+  if (ty instanceof VectorType) {
+    const v = ty;
+    return Type.vec(v.width, cty as ScalarType);
+  }
+  if (ty instanceof MatrixType) {
+    const m = ty;
+    return Type.mat(m.cols, m.rows, cty as ScalarType);
+  }
+  if (ty instanceof ArrayType) {
+    const a = ty;
+    return Type.array(a.count, cty);
+  }
+  throw new Error(`unhandled type ${ty}`);
+}
```

Comment:
nit: you're double-dispatching with `elementTypeOf()` then switching on the type. Can be reduced to:

```suggestion
/** 
 * @returns the concrete type of the given Type.
 * Example: vec3<abstract-float> -> vec3<float>
 */
export function concreteTypeOf(ty: Type): Type {
  switch (true) {
    case ty == Type.abstractInt:
      return Type.i32;
    case ty == Type.abstractFloat:
      return Type.f32;
    case ty instanceof ScalarType:
      return ty;
    case ty instanceof VectorType:
      return Type.vec(ty.width, concreteTypeOf(ty.elementType) as ScalarType);
    case ty instanceof MatrixType:
      return Type.mat(ty.cols, ty.rows, concreteTypeOf(ty.elementType) as ScalarType);
    case ty instanceof ArrayType:
      return Type.array(ty.count, concreteTypeOf(ty.elementType));
    default:
      throw new Error(`unhandled type ${ty}`);
  }
}
```

Reply:
Wow, that's a neat thing you can do in JS - match over expressions.